protocol.md

```diff
@@ -0,0 +1,196 @@
+
+This protocol consist of two components:
+* payload specification
+* Whisper adapter.
+
+The payload specification describes how the messages are encoded and decoded and what each fields mean. This is required to properly interpret messages by the client.
+
+Whisper adapter specifies interaction with the Whiper service with regards to keys management, configuration and some metadata (like topic) required to properly process and encrypt/decrypt messages.
+
+The protocol does not specify things like peers discovery, however, some notes and the current implementation will be described in an appendix.
+
+# Payload
+
+Payload is encrypted and decrypted using [transit format](https://github.com/cognitect/transit-format). It is a text-based format so that each decrypted message's payload can be easily read.
+
+Payload contains the following fields:
+* text `string`
+* content type `enum { text/plain }`
+* message type `enum { public-group-user-message, user-message }`
+* clock `int64`
+* timestamp `int64`
+* content `struct { chat-id string, text string }`
+
+Example of a valid message:
+
+```
+["~#c4",["abc123","text/plain","~:public-group-user-message",154593077368201,1545930773682,["^ ","~:chat-id","testing-adamb","~:text","abc123"]]]
+```
+
+As you can see, the message is an array and each index value has its meaning:
+* 0: `c4` is a decoder handler identificator for the current payload format. Identificators allow to register handler for many different types of messages
+* 1: array which items correspond to the described payload fields above
+
+For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
+
+## Content types
+
+Here is a list of supported content types:
+* `text/plain` identifies a message which content is a plain text
+
+TODO: are there other types? How is it useful? Sending Eth requests?
+
+## Message types
+
+Here is a list of supported message types:
+* `public-group-user-message` is a message to the public group
+* `user-message` is a private message.
+
+TODO: how is it useful? What are other supported message types?
+
+## Message ordering
+
+TBD
+
+## Quoting
+
+TBD
+
+# Whisper adapter
+
+Whisper has been chosen as an messages exchange protocol because it was designed as an off-chain communcation layer for the Ethereum nodes. It supports e2e encryption and uses epidemic spread to route data to all members of the network. It also provides [darkness to some extent](https://github.com/ethereum/go-ethereum/wiki/Achieving-Darkness).
+
+However, Whisper was not designed to handle huge number of messages and real-time communication. These are the tradeoffs that we accepted when implementing this protocol.
```

Comment:
It would be useful to point, which tradeoffs are there exactly.

Reply:
I can rewrite this part but we accepted tradeoffs that Whisper can't handle huge amount of concurrent messages and is not designed for real-time communicate, i.e. there is a latency often > 1-2s.

Reply:
yeah, I would explicitly state these ones in the doc, they are useful to know.

Reply:
Also, why these trade-offs exist fundamentally... if not here, then a link to them.

Reply:
Anonymity trilemma, though worth noting this hasn't been studied rigorously by any academics etc and it's largely a theoretical claim